netket/utils/jax.py

```diff
@@ -20,3 +20,27 @@ def get_afun_if_module(mod_or_fun, *args, **kwargs):
         return mod_or_fun.apply
     else:
         return mod_or_fun
+
+
+class WrappedApplyFun:
+    def __init__(self, module):
+        self.apply = module
+        self._hash = None
+
+    def __eq__(self, other):
+        return type(other) is WrappedApplyFun and self.apply == other.apply
+
+    def __hash__(self):
+        if self._hash is None:
+            self._hash = hash(self.apply)
+        return self._hash
```

Comment:
Could `WrappedApplyFun` be implemented as `netket.utils.struct.dataclass`? (In that case, this could use the `@property_cached` cached decorator as a small added bonus.)

Reply:
Also, docstring for the class.

However, I'm still not 100% sold about this way of passing stuff to the samplers.

Reply:
Ok, let me try it

Reply:
I think we can just use the default `__hash__` of NetKet `dataclass`, and I would not add complexity here by caching the hash. If the hash of the wrapped callable is expensive, the cache should be implemented in the callable.